### PR TITLE
View all courses

### PR DESCRIPTION
**Pull request by apalaniuk**
@d2l-joelp, @ryantmer 

There are still a couple of issues, but I don't want them to delay other testing/feedback any longer than I already have while I look at them, especially as one of them relates to a Polymer behavior:

* Keyboard navigation in the All Courses view when not using Shadow DOM seems to be problematic, preventing iron-overlay-behavior wraparound logic from working. 
* Translating isn't implemented as the library/element to use hasn't been completely decided yet. That's intentional and will be revisited.

### CODE REVIEW

**Comment by ryantmer** [comment]
I think my merge gave you some conflicts, you'll wanna rebase :(

File: d2l-all-courses-styles.html [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+	<template>
+		<style include="vui-typography"></style>
+		<style include="d2l-course-tile-region-styles"></style>
+		<style>
+		:host {
+			@apply(--vui-font);
+			display: block;
+			font-family: 'Lato-Regular', sans-serif;
+			letter-spacing: 1px;
+			color: var(--vui-color-ferrite);
```
Comment (ryantmer) [inline d2l-all-courses-styles.html]: Nice, so we can apply VUI styles now? Nice.

File: d2l-all-courses-view.html [hunk 2]
```diff
@@ -0,0 +1,253 @@
+<link rel="import" href="../polymer/polymer.html">
```
Comment (d2l-joelp) [inline d2l-all-courses-view.html]: nitpick: Probably don't need to have '-view' in this file name.

File: d2l-all-courses-view.html [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+<link rel="import" href="d2l-all-courses-styles.html">
+<link rel="import" href="d2l-simple-overlay.html">
+<link rel="import" href="d2l-course-tile.html">
+<link rel="import" href="../d2l-ajax-ui/d2l-ajax.html">
```
Comment (d2l-joelp) [inline d2l-all-courses-view.html]: If you rebase, this reference will change to `../d2l-ajax/d2l-ajax.html`

File: d2l-all-courses-view.html [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+
+		<d2l-simple-overlay
+			id='all-courses-overlay'
+			title={{allCourses}}
+			with-backdrop
+			restore-focus-on-close>
+
+			<div class="dialog-content">
+				<h3 class="header">{{pinnedCourses}}</h3>
+				<div class="my-courses-container" role="region" aria-label="Pinned Courses">
```
Comment (d2l-joelp) [inline d2l-all-courses-view.html]: Should localize the `aria-label`.

File: d2l-all-courses-view.html [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+					<template is="dom-repeat" items="[[pinnedCoursesEntities]]">
+						<d2l-course-tile
+							target="/d2l/home/[[item.properties.id]]"
```
Comment (d2l-joelp) [inline d2l-all-courses-view.html]: Since we're now specifying the url to the course in multiple places, maybe we should move it into the course tile itself and just pass in the course id? Unless we want to make the course tile ignorant of where it points to.

File: d2l-all-courses-view.html [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+							name="[[item.properties.name]]">
+						</d2l-course-tile>
+					</template>
+				</div>
+				<h3 class="header">{{unpinnedCourses}}</h3>
+				<div class="my-courses-container" role="region" aria-label="Unpinned Courses">
```
Comment (d2l-joelp) [inline d2l-all-courses-view.html]: Localize `aria-label`

File: d2l-all-courses-view.html [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+					readOnly: true
+				},
+				allEnrollmentsEntities: {
+					type: Array,
+					readOnly: true
+				},
+				enrollmentsUrl: {
+					type: String
+				},
+				windowResizeHandler: {
```
Comment (d2l-joelp) [inline d2l-all-courses-view.html]: Should probably implement this window resize handling as a behavior now that it's being used in two spots as there's a fair amount of code that was copied.

File: d2l-all-courses-view.html [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+					value: function () {
+						return this._handleWindowResize.bind(this);
+					}
+				}
+			},
+			observers: [
+				'_updateData(messages, locale)'
+			],
+			behaviors: [
+				Polymer.d2l.formatJSBehavior
```
Comment (d2l-joelp) [inline d2l-all-courses-view.html]: Let's chat about how we're doing localization :)

File: d2l-all-courses-view.html [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+					this._rescaleCourseTileRegions();
+					this.$.allEnrollmentsRequest.generateRequest();
+					this.$$('d2l-simple-overlay').open();
+			},
+			_isCourseUnpinned: function(pinnedCoursesEntities) {
+				return function(element) {
+					var isUnpinned = true;
+
+					pinnedCoursesEntities.forEach(function(pinnedCourse) {
+						if(element.properties.id == pinnedCourse.properties.id) {
```
Comment (d2l-joelp) [inline d2l-all-courses-view.html]: You can't exit a `forEach` early, but if you use a regular loop you can. Then, if you find the element you're looking for you should be able to just `return false` immediately and `return true` at the end if it wasn't found.

File: d2l-all-courses-view.html [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+				this._getTranslations();
+			},
+			_getTranslations: function() {
+				if(this.messages) {
+					this.allCourses = this.format(this.getCultureTerm('allCourses'), null);
+					this.pinnedCourses = this.format(this.getCultureTerm('pinnedCourses'), null);
+					this.unpinnedCourses = this.format(this.getCultureTerm('unpinnedCourses'), null);
+				}
+			},
+			focusableNodesOverride: function() {
```
Comment (d2l-joelp) [inline d2l-all-courses-view.html]: I will admit, I don't understand what this focusable nodes stuff is for :)

File: d2l-simple-overlay.html [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+			this.close();
+		},
+		_isMobile: function () {
+			var width = Math.max(document.documentElement.clientWidth, window.innerWidth || 0);
+			return width < 768;
+		},
+		_updateBackdropStyle: function () {
+			var style = document.createElement('style', 'custom-style');
+			style.innerHTML = "iron-overlay-backdrop { --iron-overlay-backdrop-opacity: 0.0; }";
+			document.body.appendChild(style);
```
Comment (d2l-joelp) [inline d2l-simple-overlay.html]: I think you should be able to style this normally using this style in `d2l-simple-overlay-styles`
```css
iron-overlay-backdrop {
  --iron-overlay-backdrop-opacity: 0;
}
```

Comment (apalaniuk) [inline d2l-simple-overlay.html]: I did try this at first, but because the backdrop is applied to the body element, it didn't override the default value. I'll take another look to see if I can it working in a better way again.

Comment (d2l-joelp) [inline d2l-simple-overlay.html]: Sure. It just looked from the 'docs' like that should be possible: https://github.com/PolymerElements/iron-overlay-behavior/blob/master/iron-overlay-backdrop.html#L19. I could be wrong.

File: d2l-all-courses-styles.html [hunk 12]
```diff
@@ -0,0 +1,24 @@
+<link rel="import" href="../vui-typography/typography.html">
```
Comment (ryantmer) [inline d2l-all-courses-styles.html]: I'm thinking this file shouldn't really be required - since the all courses view should have the same styling as the pinned-only view... or maybe we should just look at combining them into one file? Outside of this PR, either way. I'll make an issue for it.

File: d2l-all-courses-view.html (hunk 5 above)
Comment (ryantmer) [inline d2l-all-courses-view.html]: (mentioned earlier, but to make it all official-like) as I'm adding the API calls to the course tile, it looks like it might actually be cleaner to pass the whole enrollment Entity. So, I'll be updating this anyway, so should be fine for now.

File: d2l-all-courses-view.html [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+							isUnpinned = false;
+							return;
+						}
+					});
+
+					return isUnpinned;
+				}
+			},
+			_filterUnpinnedCourses: function(pinnedCoursesEntities, allCoursesEntities) {
+				if(pinnedCoursesEntities && allCoursesEntities) {
```
Comment (ryantmer) [inline d2l-all-courses-view.html]: Don't actually need to do a comparison with the pinned courses here; the entities have a `pinDate` property, so you can just check for that to know if they're pinned or unpinned.

File: d2l-all-courses-view.html [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+			},
+			focusableNodesOverride: function() {
+				var courseTileFocusables = [];
+
+				this.getContentChildren().forEach(function(contentChild) {
+					var courseTiles = Polymer.dom(contentChild).querySelectorAll('d2l-course-tile');
+
+					[].forEach.call(courseTiles, function(courseTile) {
+						Array.prototype.push.apply(courseTileFocusables, Polymer.dom(courseTile.root).querySelectorAll('a'));
+					})
```
Comment (ryantmer) [inline d2l-all-courses-view.html]: Missing semicolon - weird that the linter didn't get that :confused:

File: d2l-course-tile-styles.html [hunk 15]
```diff
@@ -1,22 +1,38 @@
...
+		:focus .course-text {
+			text-decoration: underline;
+		}
+
 		a, a:hover, a:visited {
 			text-decoration: none;
 			color: inherit;
 		}
+
+		a:focus .course-text {
```
Comment (ryantmer) [inline d2l-course-tile-styles.html]: If you move the `:focus .course-text` block below the `a, a:hover, a:visited` block, shouldn't need this I don't think?

File: d2l-course-tile-styles.html [hunk 16]
```diff
@@ -25,20 +41,23 @@
 			bottom: 0;
 			border-top-left-radius: 10px;
 			border-top-right-radius: 10px;
-			background: #B9C2D0;
+			background: var(--vui-color-pressicus);
```
Comment (ryantmer) [inline d2l-course-tile-styles.html]: :+1:

File: d2l-formatjs-behavior.html [hunk 17]
```diff
@@ -0,0 +1,45 @@
+<link rel="import" href="../polymer/polymer.html">
+<!--<link rel="import" href="../formatjs-element/formatjs-behavior.html">-->
```
Comment (ryantmer) [inline d2l-formatjs-behavior.html]: Is this intentionally excluded? If so, can it be removed?

File: d2l-formatjs-behavior.html [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+			locale: {
+				type: String,
+				value: 'en-US'
+			}
+		},
+		format: function(msg, values, language) {
+			language = language || this.locale;
+
+			// @TODO: Implement translation, once a suitable library/element is determined/created.
+			//return Polymer.formatJSBehavior.translate(msg, values, language);
```
Comment (ryantmer) [inline d2l-formatjs-behavior.html]: For possible-future-reference, I think this should be `Polymer.d2l`, ja?

File: d2l-all-courses-view.html (hunk 10 above)
Comment (apalaniuk) [inline d2l-all-courses-view.html]: iron-overlay-behavior tries to select focusable elements automatically (so that keyboard focus wraparound restricted to dialog elements alone works.. or is supposed to at least..), but its internal selector isn't suitable for how we're rendering course tiles, so we need to override it so it knows what the first and last children are.

I'm also not 100% sure I understand it..

Comment (d2l-joelp) [inline d2l-all-courses-view.html]: Makes enough sense for me :)

File: d2l-formatjs-behavior.html (hunk 18 above)
Comment (apalaniuk) [inline d2l-formatjs-behavior.html]: @ryantmer: Polymer.formatJSBehavior is from [formatjs-element](https://github.com/felixzapata/formatjs-element), but I should remove any traces of implementation if it's effectively unimplemented.

File: d2l-simple-overlay.html (hunk 11 above)
Comment (apalaniuk) [inline d2l-simple-overlay.html]: I tried this about every way imaginable, including by styling :host and :root from a custom Polymer element that was also child of the body element, but it doesn't seem possible to do it from a web component. I will make it a bit more idiomatic though.

File: d2l-course-tile.html [hunk 19]
```diff
@@ -9,6 +9,7 @@
 			<div
 				class="course-image-container">
 				<div class="course-image"></div>
+				<div class="course-image"></div>
```
Comment (d2l-joelp) [inline d2l-course-tile.html]: Looks like this got duplicated. Maybe a rebase issue?

File: d2l-course-tile.html [hunk 20]
```diff
@@ -60,7 +64,7 @@
 				if (newValue) {
 					Polymer.dom(courseImage).setAttribute('style', 'background-image: url(' + newValue + ');');
 				}
-			},
+			},
```
Comment (d2l-joelp) [inline d2l-course-tile.html]: nit: extraneous whitespace

File: d2l-course-tile-styles.html [hunk 21]
```diff
@@ -1,22 +1,34 @@
...
 		}
+
+		:focus .course-text {
+			text-decoration: underline;
+		}
+
 		.course-image-container {
 			position: relative;
 		}
+
```
Comment (ryantmer) [inline d2l-course-tile-styles.html]: the smallest of nits: extra line breaks

File: d2l-course-tile.html [hunk 22]
```diff
@@ -47,6 +47,9 @@
 					observer: '_pinnedChanged'
 				}
 			},
+			ready: function () {
+				this.setAttribute('role', 'link');
```
Comment (ryantmer) [inline d2l-course-tile.html]: I r confus by this - why does this have to wait until `ready` to be set? :confetti_ball:

**Comment by ryantmer** [comment]
This LGTM :+1: - any further thoughts? @d2l-joelp @apalaniuk

**Comment by apalaniuk** [comment]
@ryantmer @d2l-joelp 

Anything additional I'll put into an additional PR.
